src/NServiceBus.NHibernate/SynchronizedStorage/NHibernateAmbientTransactionSynchronizedStorageSession.cs

```diff
@@ -20,29 +21,34 @@ public NHibernateLazyAmbientTransactionSynchronizedStorageSession(Func<IDbConnec
         }
 
         public ISession Session => session.Value;
-
-        public void Dispose()
+        public void OnSaveChanges(Func<SynchronizedStorageSession, Task> callback)
         {
-            try
+            if (onSaveChangesCallback != null)
             {
-                if (session.IsValueCreated)
-                {
-                    session.Value.Flush();
-                    session.Value.Dispose();
-                }
+                throw new Exception("Save changes callback for this session has already been registered.");
             }
-            finally
+            onSaveChangesCallback = callback;
+        }
+
+        public void Dispose()
+        {
+            if (connection.IsValueCreated)
             {
-                if (connection.IsValueCreated)
-                {
-                    connection.Value.Dispose();
-                }
+                connection.Value.Dispose();
             }
         }
 
-        public Task CompleteAsync()
+        public async Task CompleteAsync()
         {
-            return Task.FromResult(0);
+            if (onSaveChangesCallback != null)
+            {
+                await onSaveChangesCallback(this).ConfigureAwait(false);
+            }
+            if (session.IsValueCreated)
```

Comment:
I think this is fine as the session is used in a `using` block and the dispose will take care of `session.Value`.